evm/src/cpu/kernel/interpreter.rs

```diff
@@ -181,6 +196,137 @@ impl<'a> Interpreter<'a> {
         result
     }
 
+    /// Initialize interpreter state given `GenerationInputs`. This assumes that the
+    /// loaded code is the KERNEL code.
+    pub(crate) fn initialize_interpreter_state_with_kernel(&mut self, inputs: GenerationInputs) {
+        self.initialize_interpreter_state(inputs, KERNEL.code_hash, KERNEL.code.len());
+    }
+
+    /// Initialize interpreter state given `GenerationInputs`.
+    pub(crate) fn initialize_interpreter_state(
+        &mut self,
+        inputs: GenerationInputs,
+        kernel_hash: H256,
+        kernel_code_len: usize,
+    ) {
+        let tries = &inputs.tries;
+
+        // Set state's inputs.
+        self.generation_state.inputs = inputs.clone();
+
+        // Initialize the MPT's pointers.
+        let (trie_root_ptrs, trie_data) =
+            load_all_mpts(tries).expect("Invalid MPT data for preinitialization");
+        let trie_roots_after = &inputs.trie_roots_after;
+        self.generation_state.trie_root_ptrs = trie_root_ptrs;
+
+        // Initialize the `TrieData` segment.
+        for (i, data) in trie_data.iter().enumerate() {
+            let trie_addr = MemoryAddress::new(0, Segment::TrieData, i);
+            self.generation_state.memory.set(trie_addr, data.into());
+        }
+
+        // Update the RLP and withdrawal prover inputs.
+        let rlp_prover_inputs =
+            all_rlp_prover_inputs_reversed(inputs.clone().signed_txn.as_ref().unwrap_or(&vec![]));
+        let withdrawal_prover_inputs = all_withdrawals_prover_inputs_reversed(&inputs.withdrawals);
+        self.generation_state.rlp_prover_inputs = rlp_prover_inputs;
+        self.generation_state.withdrawal_prover_inputs = withdrawal_prover_inputs;
+
+        // Set `GlobalMetadata` values.
+        let metadata = &inputs.block_metadata;
+        let global_metadata_to_set = [
+            (
+                GlobalMetadata::BlockBeneficiary,
+                U256::from_big_endian(&metadata.block_beneficiary.0),
+            ),
```

Comment:
probably out of scope for this PR, but we're starting to have quite some duplication between the interpreter and the actual prover, so we may want to start unifying everything. Would be easier to maintain as well.